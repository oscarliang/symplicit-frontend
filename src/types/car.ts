export interface Car {
  _id: string;
  brand: string;
  drive: string;
  id: number;
  imageUrl: string;
  name: string;
  price: string;
}

export interface AllCars {
  allCars: Car[];
}

export interface Modules {
  modules: Record<string, number>;
}

export interface GraphqlResponse<T = never> {
  data: {
    data: T;
    errors: unknown;
  };
}

export interface AllCarResponse extends GraphqlResponse<AllCars> {}

export interface RestfulResponse<T = never> {
  data: {
    data: T;
    modules: Record<string, number>;
  };
}

export interface AllCarRestFulResponse extends RestfulResponse<AllCars> {}

export interface ModulesRestFulResponse extends RestfulResponse<Modules> {}

export interface GetCarByNameResponse {
  data: {
    data: {
      getCarByName: Car[];
    };
  };
}

export interface GetCarByBrandResponse {
  data: {
    data: {
      getCarByBrand: Car[];
    };
  };
}

export interface GetCarByDriveResponse {
  data: {
    data: {
      getCarByDrive: Car[];
    };
  };
}

export interface CreateCarResponse {
  data: {
    data: {
      createCar: Car;
    };
  };
}

export interface DeleteCarResponse {
  data: {
    data: {
      deleteCar: Car;
    };
  };
}

export interface CarRequest {
  _id?: string;
  brand: string;
  drive: string;
  id?: number;
  imageUrl: string;
  name: string;
  price: string;
}

export interface NormalizedCarsEntity {
  entities?: {
    cars: Car[];
  };
  result?: number[];
}

export interface NormalizedEntity {
  cars: NormalizedCarsEntity;
}
