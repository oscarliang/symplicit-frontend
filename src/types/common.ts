export interface ApiError {
  message: string;
  path: string;
  statusCode: number;
  timestamp: string;
}

export interface RestApiResponse<T> {
  data: {
    statusCode: number;
    message: string;
    data: T;
  };
}
