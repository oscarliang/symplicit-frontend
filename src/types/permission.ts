export interface Permission {
  module: string;
  permission: number;
}

export enum PermissionType {
  CREATE = 1,
  READ = 2,
  UPDATE = 4,
  DELETE = 8,
}
