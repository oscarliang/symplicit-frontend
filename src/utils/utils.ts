export function isObject(object: unknown): object is Record<string, unknown> {
  return !!object && !Array.isArray(object) && typeof object === 'object';
}

/* eslint-disable no-param-reassign */
export function merge(...objects: unknown[]): unknown {
  return objects.reduce((current: Record<string, unknown>, object) => {
    if (!isObject(object)) return current;

    Object.keys(object).forEach((key) => {
      const currentValue = current[key];
      const objectValue = object[key];

      if (isObject(currentValue) && isObject(objectValue)) {
        current[key] = merge(currentValue, objectValue);
      } else if (Array.isArray(objectValue)) {
        current[key] = objectValue.slice(0);
      } else {
        current[key] = objectValue;
      }
    });

    return current;
  }, {});
}
/* eslint-enable no-param-reassign */

export function cssStylesToObject(styles: string): Record<string, string> {
  const reactStyles: Record<string, string> = {};
  styles.split(';').forEach((style) => {
    const [key, value] = style.split(':').map((s) => (s ? s.trim() : s));
    if (!key || !value) return;
    const camelcaseKey = key.replace(/(-\w)/g, (word: string) =>
      word[1].toUpperCase(),
    );
    reactStyles[camelcaseKey] = value;
  });
  return reactStyles;
}

export function snakeToCamel(value: string): string {
  return value
    .replace(/([a-z])([A-Z])/g, '$1_$2')
    .split(/_/g)
    .map((segment, idx) => {
      const firstLetter = segment.slice(0, 1);
      const rest = segment.slice(1);
      return `${firstLetter[
        idx === 0 ? 'toLowerCase' : 'toUpperCase'
      ]()}${rest.toLowerCase()}`;
    })
    .join('');
}

/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

const excludedKeys = ['_id'];

export function keysToCamel(obj: any): any {
  if (!obj || typeof obj !== 'object') {
    return obj;
  }

  if (Array.isArray(obj)) {
    return obj.map((item) => keysToCamel(item));
  }

  const json: Record<string, any> = obj as Record<string, any>;
  const newObj: Record<string, any> = {};
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  Object.keys(obj).forEach((key) => {
    if (excludedKeys.includes(key)) {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      newObj[key] = json[key];
    } else {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      newObj[snakeToCamel(key)] = keysToCamel(json[key]);
    }
  });

  return newObj;
}
/* eslint-enable @typescript-eslint/no-unsafe-return */
/* eslint-enable @typescript-eslint/no-explicit-any */

export function camelToSnake(string: string, splitOnNumbers = true): string {
  return string.replace(
    splitOnNumbers ? /[A-Z0-9]/g : /[A-Z]/g,
    (letter) => `_${letter.toLowerCase()}`,
  );
}

export function camelToWords(string: string): string {
  return string.replace(/[A-Z]/g, (letter) => ` ${letter.toLowerCase()}`);
}

export function wordToCamel(value: string): string {
  return value
    .trim()
    .split(/ /g)
    .map((segment, idx) => {
      const firstLetter = segment.slice(0, 1);
      const rest = segment.slice(1);
      return `${firstLetter[
        idx === 0 ? 'toLowerCase' : 'toUpperCase'
      ]()}${rest.toLowerCase()}`;
    })
    .join('');
}

export function camelKeysToSnake(
  obj: Record<string, unknown>,
  splitOnNumbers = true,
): Record<string, unknown> {
  return Object.entries(obj).reduce(
    (newObj, [key, value]) => ({
      ...newObj,
      [camelToSnake(key, splitOnNumbers)]: value,
    }),
    {},
  );
}

export function nameToInitials(name: string): string {
  // Basic function to create initials from a name.
  // - Not smart to detect if index = 1 is middlename or lastname
  // - If name is one word, it uses two first letters of that single word
  const nameParts = name.replace(/\t/g, ' ').trim().split(' ');
  return nameParts.length === 1
    ? nameParts[0].split('').slice(0, 2).join('') || ''
    : `${nameParts[0][0]}${nameParts.slice(-1)[0][0]}`;
}

function titleCaseWord(word: string): string {
  if (word.toLowerCase() === 's ') {
    // Assumes this is the single "s" after an "'s",
    // so want to keep it lowercase.
    return word.toLowerCase();
  }
  return word.charAt(0).toUpperCase() + word.substring(1).toLowerCase();
}

export function titleCase(name: string | null): string {
  if (!name) return '';

  // Transform to title case
  const basicCapitalize = name
    .split(' ')
    .map(
      (part) => `${part.charAt(0).toUpperCase()}${part.slice(1).toLowerCase()}`,
    )
    .join(' ');

  return basicCapitalize
    .replace(/[^\s\-']+[\s\-']*/g, (word: string) => titleCaseWord(word))
    .replace(/\b(Van|De|Der|Da|Von)\b/g, (nobiliaryParticle) =>
      nobiliaryParticle.toLowerCase(),
    )
    .replace(/\b(And|To|The|From|Of|At|For|By|On|With|In)\b/g, (specialCase) =>
      // Some special cases:
      // - `TO`, `THE` and `EDITOR` LETTERS TO THE EDITOR
      // - `AND` in case `AUTHOR1 AND AUTHOR2`
      specialCase.toLowerCase(),
    )
    .replace(
      /Mc(.)/g,
      (_match, letter3: string) => `Mc${letter3.toUpperCase()}`,
    )
    .replace(/^(\w{1})/g, (firstLetter: string) => firstLetter.toUpperCase());
}
