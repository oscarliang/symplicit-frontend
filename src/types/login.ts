import { RestApiResponse } from './common';

export interface User {
  password: string;
  username: string;
}

export interface Login {
  accessToken: string;
  modules: Record<string, number>[];
}

export interface LoginResponse extends RestApiResponse<Login> {}
