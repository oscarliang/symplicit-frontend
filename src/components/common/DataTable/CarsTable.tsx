import * as React from 'react';
import { useState } from 'react';
import { NumericFormat } from 'react-number-format';
import {
  Card,
  CardBody,
  CardHeader,
  Col,
  Pagination,
  PaginationItem,
  PaginationLink,
  Row,
  Table,
} from 'reactstrap';

import { removeCarAction } from '../../../ducks/cars';
import { SelectCarType, updateMessage } from '../../../ducks/global';
import { deleteCarService, selectCar } from '../../../service/car-service';
import { useAppDispatch } from '../../../store';
import { Car, CarRequest } from '../../../types/car';
import Confirm from '../Confirm/Confirm';

export interface CarsTableProps {
  allCars: Record<number, CarRequest>; // inherit from the home-page=containers
}

function CarsTable({ allCars }: CarsTableProps): React.ReactElement {
  const [isActive, setIsActive] = useState(-1);
  const dispatch = useAppDispatch();

  const [selectDeleteCar, setSelectDeleteCar] = useState<CarRequest | null>(
    null,
  );
  const [open, setOpen] = useState(false);
  const [currentPage, setCurrentPage] = useState(0);
  const [confirmTitle, setConfirmTitle] = useState('');
  const [confirmMessage, setConfirmMessage] = useState('');
  const pageSize = 6;

  const onClickConfirmDelete = (car: CarRequest) => {
    setOpen(true);
    setConfirmTitle('Warning');
    setConfirmMessage(`Are you sure to delete ${car.name}?`);
    setSelectDeleteCar(car);
  };

  const handlePageClick = (e: React.MouseEvent, index: number) => {
    e.preventDefault();
    setCurrentPage(index);
  };

  const onClickCloseConfirm = () => {
    setOpen(false);
  };

  const onClickYesButton = () => {
    setOpen(false);
    if (selectDeleteCar) {
      deleteCarService(selectDeleteCar)
        .then((carObj: Car) => {
          dispatch(removeCarAction(carObj));
          dispatch(updateMessage(`${carObj.name} has been deleted`));
        })
        .catch((err) => {
          console.error(err);
          dispatch(
            updateMessage(
              'Oops! It seems to be some error. Please contact the admin',
            ),
          );
        });
    }
  };

  const onClickCar = (index: number, car: SelectCarType) => {
    dispatch(selectCar(car));

    // update the select row
    setIsActive(index);
  };

  const renderTableBody = () => {
    const cars = Object.values(allCars) as unknown as SelectCarType[];
    return cars.slice(currentPage * pageSize, (currentPage + 1) * pageSize).map(
      (car: SelectCarType, index: number) =>
        car && (
          <tr
            className={isActive === index ? 'table-info' : ''}
            data-testid={`car-${index}`}
            key={car.id}
            onClick={() => onClickCar(index, car)}
          >
            <td>
              <img
                alt={car.name}
                className='rounded car-img animated bounceIn'
                src={car.imageUrl}
              />
            </td>
            <td>{car.name}</td>
            <td>{car.brand.toUpperCase()}</td>
            <td>{car.drive.replace(/\|/g, ' ').toUpperCase()}</td>
            <td aria-label='price'>
              <NumericFormat
                displayType='text'
                prefix='$'
                thousandSeparator
                value={car.price}
              />
            </td>
            <td
              aria-label='delete button'
              onClick={(e: React.MouseEvent) => {
                e.preventDefault();
                e.stopPropagation();
                onClickConfirmDelete(car);
              }}
            >
              <i className='fa fa-times' />
            </td>
          </tr>
        ),
    );
  };

  return (
    <Col>
      <Row>
        <Card>
          <CardHeader>
            <i className='fa fa-align-justify' /> All Cars
          </CardHeader>
          <CardBody>
            <Table bordered hover responsive size='sm' striped>
              <thead>
                <tr>
                  <th>Image</th>
                  <th>Name</th>
                  <th>Brand</th>
                  <th>Model</th>
                  <th>Price</th>
                  <th>Delete</th>
                </tr>
              </thead>
              <tbody>{renderTableBody()}</tbody>
            </Table>
          </CardBody>
        </Card>
        <Confirm
          body={confirmMessage}
          isOpen={open}
          onClickClose={onClickCloseConfirm}
          onClickYesButton={onClickYesButton}
          title={confirmTitle}
        />
      </Row>
      <Row>
        <Col>
          <Pagination>
            {Array.from({
              length: Math.floor(Object.values(allCars).length / pageSize) + 1,
            }).map((page, i) => (
              // eslint-disable-next-line react/no-array-index-key
              <PaginationItem active={i === currentPage} key={i}>
                <PaginationLink href='#' onClick={(e) => handlePageClick(e, i)}>
                  {i + 1}
                </PaginationLink>
              </PaginationItem>
            ))}
          </Pagination>
        </Col>
      </Row>
    </Col>
  );
}

export default CarsTable;
