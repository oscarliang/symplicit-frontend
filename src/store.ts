import { PayloadAction, configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { loadingBarReducer } from 'react-redux-loading-bar';
import { combineReducers } from 'redux';
import * as Redux from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';

import carsReducer from './ducks/cars';
import globalReducer from './ducks/global';
import LoginReducer, { LOG_OUT_ACTION } from './ducks/login';
import permissionReducer from './ducks/permission';
import { merge } from './utils/utils';

const ACTION_HYDRATE = 'HYDRATE';
export function hydrateAction(
  state: Partial<RootState>,
): PayloadAction<Partial<RootState>> {
  return {
    payload: state as RootState,
    type: ACTION_HYDRATE,
  };
}

const combinedReducer = combineReducers({
  authenticated: LoginReducer,
  cars: carsReducer,
  global: globalReducer,
  loadingBar: loadingBarReducer,
  permissions: permissionReducer,
});

export function createStore(configure?: (state: RootState) => RootState) {
  const middleware: [Redux.Middleware] = [thunk];

  if (
    process.env.NODE_ENV !== 'production' &&
    process.env.NEXT_PUBLIC_DEBUGGER
  ) {
    // add the previous state and next state info
    middleware.push(
      createLogger({
        collapsed: true,
        level: 'info',
      }),
    );
  }

  const store = configureStore({
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(middleware),
    reducer: (
      state: RootState | undefined,
      action: PayloadAction<Partial<RootState>>,
    ) => {
      // Clear all data in redux store to initial.
      if (action.type === LOG_OUT_ACTION) {
        // eslint-disable-next-line no-param-reassign
        state = undefined;
      }

      if (action.type === ACTION_HYDRATE) {
        return merge(state, action.payload) as RootState;
      }
      return combinedReducer(state, action);
    },
  });

  if (configure) {
    store.dispatch(hydrateAction(configure(store.getState())));
  }

  return store;
}

export const store = createStore();

export type RootState = ReturnType<typeof combinedReducer>;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
