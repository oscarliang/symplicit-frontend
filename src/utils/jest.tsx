import { EnhancedStore } from '@reduxjs/toolkit';
import * as React from 'react';
import { Provider } from 'react-redux';
import { SWRConfig } from 'swr';

import { RootState } from 'store';

interface TestWrapperProps {
  children: React.ReactNode;
  store: EnhancedStore<RootState>;
}

// eslint-disable-next-line import/prefer-default-export
export function TestWrapper({
  children,
  store,
}: TestWrapperProps): React.ReactElement {
  return (
    <SWRConfig value={{ provider: () => new Map() }}>
      <Provider store={store}>{children}</Provider>
    </SWRConfig>
  );
}
