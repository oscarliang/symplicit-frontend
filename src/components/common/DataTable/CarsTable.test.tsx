import { render, screen } from '@testing-library/react';
import * as React from 'react';

import AllCarsFixture from 'mocks/allCarsFixture';
import { createStore } from 'store';
import { TestWrapper } from 'utils/jest';

import { CarRequest } from '../../../types/car';

import CarsTable from './CarsTable';

describe('<CarsTable />', () => {
  it('renders', () => {
    expect.assertions(2);

    const store = createStore();

    const { container } = render(
      <TestWrapper store={store}>
        <CarsTable
          allCars={AllCarsFixture as unknown as Record<number, CarRequest>}
        />
      </TestWrapper>,
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByTestId('car-0')).toHaveTextContent('$34,324');
  });
});
