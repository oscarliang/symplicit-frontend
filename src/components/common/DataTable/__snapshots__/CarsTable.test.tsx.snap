// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarsTable /> renders 1`] = `
<div
  class="col"
>
  <div>
    <div
      class="card"
    >
      <div
        class="card-header"
      >
        <i
          class="fa fa-align-justify"
        />
         All Cars
      </div>
      <div
        class="card-body"
      >
        <div
          class="table-responsive"
        >
          <table
            class="table table-sm table-bordered table-striped table-hover"
          >
            <thead>
              <tr>
                <th>
                  Image
                </th>
                <th>
                  Name
                </th>
                <th>
                  Brand
                </th>
                <th>
                  Model
                </th>
                <th>
                  Price
                </th>
                <th>
                  Delete
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class=""
                data-testid="car-0"
              >
                <td>
                  <img
                    alt="test1"
                    class="rounded car-img animated bounceIn"
                    src="https://images.carexpert.com.au/app/uploads/2023/09/BMW-i7-M70-xDrive_HERO-16x9-2.jpg"
                  />
                </td>
                <td>
                  test1
                </td>
                <td>
                  PORSCHE
                </td>
                <td>
                  2WD 4WD AWD 
                </td>
                <td
                  aria-label="price"
                >
                  <span>
                    $34,324
                  </span>
                </td>
                <td
                  aria-label="delete button"
                >
                  <i
                    class="fa fa-times"
                  />
                </td>
              </tr>
              <tr
                class=""
                data-testid="car-1"
              >
                <td>
                  <img
                    alt="test2"
                    class="rounded car-img animated bounceIn"
                    src="https://editorial.pxcrush.net/carsales/general/editorial/porsche-911-gts-0182.jpg?width=1024&height=683"
                  />
                </td>
                <td>
                  test2
                </td>
                <td>
                  PORSCHE
                </td>
                <td>
                  2WD 4WD AWD 
                </td>
                <td
                  aria-label="price"
                >
                  <span>
                    $4,545
                  </span>
                </td>
                <td
                  aria-label="delete button"
                >
                  <i
                    class="fa fa-times"
                  />
                </td>
              </tr>
              <tr
                class=""
                data-testid="car-2"
              >
                <td>
                  <img
                    alt="test3"
                    class="rounded car-img animated bounceIn"
                    src="https://images.carexpert.com.au/app/uploads/2023/09/BMW-i7-M70-xDrive_HERO-16x9-2.jpg"
                  />
                </td>
                <td>
                  test3
                </td>
                <td>
                  BMW
                </td>
                <td>
                  2WD 4WD 
                </td>
                <td
                  aria-label="price"
                >
                  <span>
                    $999,921,212
                  </span>
                </td>
                <td
                  aria-label="delete button"
                >
                  <i
                    class="fa fa-times"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <nav
        aria-label="pagination"
        class=""
      >
        <ul
          class="pagination"
        >
          <li
            class="page-item active"
          >
            <a
              class="page-link"
              href="#"
            >
              1
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
